neural/run_workflow.py

```diff
@@ -614,3 +631,159 @@ def test_run(q_docpartitions, q_fold_config_map, bertmodel, train_val_dir, test_
+    for i_batch, samples_batch in enumerate(data_loader):
+
+        logprob_scores = []
+
+        docs_batch, docs_len, docs_sents_len, docs_attn_mask, docs_labels, docs_id = samples_batch
+
+        docs_batch = docs_batch.to(device)
+        docs_attn_mask = docs_attn_mask.to(device)
+        docs_sents_len = docs_sents_len.type(torch.int64).numpy()  # to feed this in RNN
+        docs_labels = docs_labels.type(torch.int64).to(device)
+
+        with torch.set_grad_enabled(dsettype == 'train'):
+            # print("number of examples in batch:", docs_batch.size(0))
+            num_docs_perbatch = docs_batch.size(0)
+            for doc_indx in range(num_docs_perbatch):
+                # print('doc_indx', doc_indx)
+                doc_id = docs_id[doc_indx].item()
+                if(doc_id in bert_proc_docs):
+                    # due to GPU limit
+                    embed_sents = ReaderWriter.read_tensor(bert_proc_docs[doc_id], device)
+                else:
+                    embed_sents = bert_encoder(docs_batch[doc_indx], docs_attn_mask[doc_indx],
+                                               docs_len[doc_indx].item())
+                    # add embedding to dict
+                    embed_fpath = os.path.join(sents_embed_dir, '{}.pkl'.format(doc_id))
+                    ReaderWriter.dump_tensor(embed_sents, embed_fpath)
+                    bert_proc_docs[doc_id] = embed_fpath
+
+                sents_rnn_hidden = sent_encoder(embed_sents, docs_sents_len[doc_indx],
+                                                docs_len[doc_indx].item())
+
+                # # remove the embedding from GPU
+                # bert_proc_docs[doc_id].to(cpu_device)
+                # print('sents_rnn_hidden', sents_rnn_hidden.shape)
+                enc_sents = sents_rnn_hidden
+                # print('enc_sents', enc_sents.shape)
+                doc_out, doc_attn_weights = doc_encoder(enc_sents)
+                # print('doc_out', doc_out.shape)
+                # print('doc_attn_weights', doc_attn_weights.shape)
+                # tracking attention weight for validation and test examples
+                if(dsettype in docid_attnweights_map):
+                    docid_attnweights_map[dsettype][doc_id] = doc_attn_weights
+
+                logsoftmax_scores = doc_categ_scorer(doc_out)
+                __, pred_classindx = torch.max(logsoftmax_scores, 1)  # apply max on row level
+
+                pred_class.append(pred_classindx.item())
+
+                logprob_scores.append(logsoftmax_scores)
+                sample_counter += 1
+
+            # log probabilities
+            b_logprob_scores = torch.cat(logprob_scores, dim=0)
+
+            # do some cleaning -- get more GPU ;)
+            del docs_batch, docs_len, docs_sents_len, docs_attn_mask, docs_labels, docs_id
+
+    # end of epoch
+    return {'pred_class': pred_class,
+            'logprob_score_class0': [t.data.cpu().numpy()[0] for t in b_logprob_scores],
```

Comment:
ah, I would prefer to get the elements from the tensor this way `[t[0].item() for t in b_logprob_scores.cpu()]`, the same for `class 1`

Reply:
or a one liner using `list(x.cpu().data[:, 0].numpy())`